### DISCUSSION TITLE
Rapid mutations

### DISCUSSION THREAD

**Discussion by jfrancos** [urql-graphql/urql#1840] [Q&A] [answered]
Hi,

I'm relatively new to this, working on a project template including basic user document CRUD (the goal here is a skeleton for an offline-first PWA), and trying to compensate for some issues I see when making multiple mutations in rapid succession (i.e. making a second mutation before the response to a previous mutation arrives) (using graphcache and FaunaDB, and assuming I should have everything working properly before I work on incorporating `offlineExchange`):

- With rapid updates:

  Responses from the server will often come out of order, so if I'm updating a text field (whose value comes from an associated urql query) and quickly type 'asdf' (which I can do as long as optimistic responses are set up), the 'asdf' (from the optimistic response) will get replaced with e.g. 'as' (from the out-of-order response from the server).

  For now I've solved this with a `useEffect` hook (plus associated logic) that queues updates whenever `fetching` is true (code below), but I'm wondering if there's a better way around this (especially since I then wind up not using urql's optimistic response here), or if there's already an exchange that will automatically queue mutations while there's already an in-progress mutation?  (The queue is also helpful since even if responses arrived in the correct order, with enough simultaneous outstanding mutations, FaunaDB would eventually respond with errors)

- With rapidly creating and then deleting:

  I have an optimistic response for creating a document that includes an ID obtained via `Date.now()`.  If I try to delete a new document before getting the server's response to the create mutation, it initially looks correct because of the optimistic response I've set up for the delete operation, but since the server is receiving the delete mutation with the fake ID, the response does not have the document deleted.  Does urql have a standard way of dealing with this situation?  If not, then by default I'll deal with it in my react code as with the rapid-updates issue above, but then I will have removed all of my optimistic response code from graphcache, which seems counterintuitive.

Thanks, any feedback appreciated!!

Code for queueing rapid updates:
```
  // Read
  const [
    {
      data: {
        findUserByID: {
          docs: { data: _data },
        },
      } = { findUserByID: { docs: { data: [] } } },
      readFetching,
      readError,
    },
  ] = useQuery({
    query: FindUserByID,
  });

  // Update
  const [updateDocResult, _updateDoc] = useMutation(UpdateDoc);
  const [queuedUpdates, setQueuedUpdates] = useState({});
  const updateDoc = ({ id, ...attrs }) =>
    setQueuedUpdates((queued) => ({
      ...queued,
      [id]: { ...queued[id], ...attrs },
    }));

  /*  Sync decoupled data  */
  const { fetching: updateFetching } = updateDocResult;
  useEffect(() => {
    const entries = Object.entries(queuedUpdates);
    if (updateFetching || entries.length === 0) {
      return;
    }
    const newQueue = {};
    entries.forEach(([key, value]) => {
      const serverData = _data.find((item) => item._id === key);
      const synchronized = Object.keys(value).every(
        (item) => value[item] === serverData[item],
      );
      if (!synchronized) {
        _updateDoc({
          id: key,
          ...value,
        });
        newQueue[key] = value;
      }
    });
    setQueuedUpdates(newQueue);
  }, [queuedUpdates, updateFetching]);
  const data = _data.map((item) => ({
    ...item,
    ...queuedUpdates[item._id],
  }));
```

**Comment by jfrancos** [comment]
Hi @JoviDeCroock, 

Is there any additional info I can add, in order to get feedback on this?

Best,
Justin

**Comment by JoviDeCroock** [comment, marked as answer]
Sorry about the delay, I've had an eye on this but not had an open spot to give a good elaborate answer on this.

First and foremost, welcome to the urql community :D

The successive mutations issue is one that we've seen before and honestly personally my answer to this one is, don't fire mutations in rapid succession 😓 I know this sounds kinda dumb but essentially what you're doing is firing a mutation on every keystroke which means that you have 4 "close-to-concurrent" mutations going on.
In graphCache we'll hold an "optimistic" layer for your mutation but this will be cleared when a response comes in, we could say something about the amount of optimistic layers (in this case asdf 4) and in which order they'd have to be cleared but that's a whole lot of complexity (I'm going to look into why a permanent result for 2 clears the layer of 4 though, this could be a bug on our side).
Your solution sounds correct to me though where you defer the operation until the user has a (semi-)final input.

The second problem isn't really one we've seen before, generally my advise would be to not touch optimistic entities as it's very hard to guarantee consistency without completely mimicking the behavior your server will exhibit, the behavior that we're seeing here is basically:
- create entity with temp id
- we click delete on the optimistic entity of the above
- the delete mutation returns an error

This interaction in itself makes a lot of sense as the server errors we clear the optimistic layer because it's invalid but the optimstic layer was actually affecting another optimstic layer of a currently in-flight mutation (why has this not responded?). It gets fairly complex really quick if we start mixing the optimistic behavior with real behavior 😅

We have to keep in mind that optimistic serves as an intermediate UI and it can introduce a lot of complexity when we keep work ongoing on all these temporary entities.

EDIT: I can't seem to reproduce the behavior you're describing where updates go out of band (point 1)

- client: https://codesandbox.io/s/clever-aryabhata-89vj4?file=/src/index.js
- server: https://codesandbox.io/s/optimistic-nobel-o342r?file=/schema.js:681-709
